File: frontend/src/api/desensitize.ts
import request from './index'

export interface DesensitizeRule {
  id: number | null
  name: string
  pattern: string
  mask_strategy: string
  replacement?: string
  is_builtin: boolean
  created_at?: string
}

export interface DesensitizeRuleCreate {
  name: string
  pattern: string
  mask_strategy: string
  replacement?: string
}

export function listRules(): Promise<DesensitizeRule[]> {
  return request.get('/desensitize-rules')
}

export function createRule(data: DesensitizeRuleCreate): Promise<DesensitizeRule> {
  return request.post('/desensitize-rules', data)
}

export function updateRule(id: number, data: Partial<DesensitizeRuleCreate>): Promise<DesensitizeRule> {
  return request.put(`/desensitize-rules/${id}`, data)
}

export function deleteRule(id: number): Promise<void> {
  return request.delete(`/desensitize-rules/${id}`)
}

export function assignColumnRule(columnId: number, ruleName: string | null): Promise<void> {
  return request.put(`/desensitize-rules/columns/${columnId}/rule`, { rule_name: ruleName })
}

File: backend/app/schemas/desensitize.py
from datetime import datetime

from pydantic import BaseModel, Field


class DesensitizeRuleCreate(BaseModel):
    name: str = Field(..., max_length=100)
    pattern: str
    mask_strategy: str = Field(..., max_length=50)
    replacement: str | None = None


class DesensitizeRuleUpdate(BaseModel):
    name: str | None = Field(None, max_length=100)
    pattern: str | None = None
    mask_strategy: str | None = Field(None, max_length=50)
    replacement: str | None = None


class DesensitizeRuleResponse(BaseModel):
    id: int | None = None
    name: str
    pattern: str
    mask_strategy: str
    replacement: str | None = None
    is_builtin: bool = False
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class ColumnRuleAssign(BaseModel):
    rule_name: str | None = None
